### DISCUSSION TITLE
Optical Flow Module

### DISCUSSION THREAD

**Discussion by ChristophReich1996** [kornia/kornia#2110] [Ideas]
Hi Kornia team,

I would be interested in adding some optical flow functions. I have seen that there have been a few efforts made to implement some optical flow approaches (eg. #1465). However, before implementing some of the classical approaches (Lucas–Kanade etc.), adding some utilities, such as [optical flow to rgb](https://github.com/tomrunia/OpticalFlow_Visualization), interpolating flow, occlusion estimation, or [full image warping](https://openaccess.thecvf.com/content/CVPR2021/papers/Stone_SMURF_Self-Teaching_Multi-Frame_Unsupervised_RAFT_With_Full-Image_Warping_CVPR_2021_paper.pdf) (etc.) might be interesting and useful. I would be keen to implement some of the basic flow functionalities. Let me know what you think!

**Comment by shijianjian** [comment]
Please go ahead. I think it would be a great benefit to have it. It would be the best to make them differentiable, along with some classic use cases, deep learning models, etc.

Later on, we may wrap those methods to plug-and-use modules to ease the video understand applications like I3D or something more fancy stuff.

**Comment by edgarriba** [comment]
Definitely this is inside the project scope. Can we list what exact functionality we want to include so that we can organise in different issues?

**Comment by ChristophReich1996** [comment]
@edgarriba yes sure, so my initial list of functions to be implemented would look like this:

1. Backward warp with flow (just offsets not total pixel coordinates). This would just be a wrapper around `kornia.geometry.transform.remap`.
2. Optical flow interpolation (just interpolation and scaling).
3. Flow to RGB mapping. An implementation of the classical flow visualization proposed by [Baker et al.](https://vision.middlebury.edu/flow/floweval-ijcv2011.pdf) Reference implementations are available [here](https://github.com/tomrunia/OpticalFlow_Visualization) and [here](https://github.com/ChristophReich1996/Optical-Flow-Visualization-PyTorch).
4. Standard forward-backward consistency occlusion estimation. Reference available [here](https://lmb.informatik.uni-freiburg.de/people/brox/pub/sundaram_eccv10.pdf) and [here](https://arxiv.org/pdf/2006.04902.pdf).
5. Range-map-based occlusion estimation. Reference available [here](https://arxiv.org/pdf/2006.04902.pdf).
6. Full image warping. An advanced approach for unsupervised learning of optical flow. Reference available [here](https://openaccess.thecvf.com/content/CVPR2021/papers/Stone_SMURF_Self-Teaching_Multi-Frame_Unsupervised_RAFT_With_Full-Image_Warping_CVPR_2021_paper.pdf).

From what I've seen, most flow papers are implementing these functions themselves. Having a unified implementation would facilitate easier usage of these "standard" techniques in new projects.
Beyond this, adding support for spatially augmenting flow maps would also be interesting. Additionally, implementing a common smoothness regularizer (typically used in unsupervised flow estimation) might be also interesting.

**Reply by edgarriba** [reply]
This sounds amazing to me
